aten/src/ATen/native/cuda/DepthwiseConv3d.cu

```diff
@@ -0,0 +1,655 @@
+
+      for (int k_frame = 0; k_frame < kT; ++k_frame) {
+        const int out_frame_raw = out_frame_end - k_frame * dilationT;
+        const int out_frame = out_frame_raw / dT;
+        for (int k_row = 0; k_row < kH; ++k_row) {
+          const int out_row_raw = out_row_end - k_row * dilationH;
+          const int out_row = out_row_raw / dH;
+          for (int k_col = 0; k_col < kW; ++k_col) {
+            const scalar_t op1 = __ldg(kernel_ptr++);
+            const int out_col_raw = out_col_end - k_col * dilationW;
+            const int out_col = out_col_raw / dW;
+
+            const int out_offs = (out_frame * oH + out_row) * oW + out_col;
+
+            scalar_t op2 = (scalar_t)0;
+            if (out_col >= 0 && out_row >= 0 && out_frame >= 0 &&
+                out_col < oW && out_row < oH && out_frame < oT) {
+              op2 = __ldg(gout_ptr + out_offs);
+            }
+            if (out_frame * dT == out_frame_raw &&
+                out_row * dH == out_row_raw &&
+                out_col * dW == out_col_raw) {
+              sum += op1 * op2;
+            }
+          }
+        }
+      }
+    }
+
+    grad_input[batch][in_channel][in_frame][in_row][in_col] = sum;
+  }
+}
+
+template <typename scalar_t,
+         int kKnownStrideH, int kKnownStrideW>
+__global__ void 
+conv_depthwise3d_cuda_backward_weight_kernel(
+    const PackedTensorAccessor32<scalar_t, 5> grad_output,
+    const PackedTensorAccessor32<scalar_t, 5> input,
+    PackedTensorAccessor32<scalar_t, 5> grad_kernel,
+    int dT, int dH_, int dW_,
+    int pT, int pH, int pW, 
+    int dilationT, int dilationH, int dilationW) {
+  const int kC = grad_kernel.size(0);
+  const int kT = grad_kernel.size(2);
+  const int kH = grad_kernel.size(3);
+  const int kW = grad_kernel.size(4);
+  
+  const int dH = kKnownStrideH > 0 ? kKnownStrideH : dH_;
+  const int dW = kKnownStrideW > 0 ? kKnownStrideW : dW_;
+
+  const int k_col = blockIdx.x % kW;
+  const int k_row = (blockIdx.x / kW) % kH;
+  const int k_frame = (blockIdx.x / kW / kH) % kT;
+  const int k_channel = blockIdx.x / kW / kH / kT;
+  scalar_t *result = &grad_kernel[k_channel][0][k_frame][k_row][k_col];
+
+  const int oT = grad_output.size(2);
+  const int oH = grad_output.size(3);
+  const int oW = grad_output.size(4);
+  const int iT = input.size(2);
+  const int iH = input.size(3);
+  const int iW = input.size(4);
+  const int channel_multiplier = grad_output.size(1) / input.size(1);
+  const int in_channel = k_channel / channel_multiplier;
+
+  extern __shared__ int sdata_raw[];
+  scalar_t* sdata = reinterpret_cast<scalar_t*>(sdata_raw);
+
+  if (k_channel >= kC) {
+    return;
+  }
+
+  const int laneid = threadIdx.x % C10_WARP_SIZE;
+  const int warpid = threadIdx.x / C10_WARP_SIZE;
+  const int nwarps = blockDim.x / C10_WARP_SIZE;
+
+  scalar_t grad = (scalar_t)0;
+  int batch = warpid / oT;
+  int gout_frame = warpid - batch * oT;
+  for (int outer_pos = warpid; outer_pos < input.size(0) * oT; 
```

Comment:
a comment on the strategy is useful here (warp accumulates over 2d image and loops over batches + 3rd dimension)